IntegrationTests/tests_04_performance/test_01_resources/shared.swift

```diff
@@ -136,3 +136,133 @@ func withAutoReleasePool<T>(_ execute: () throws -> T) rethrows -> T {
+            sendSomeDataIfDesiredOrClose(context: context)
+        }
+        
+        private func sendSomeDataIfDesiredOrClose(context: ChannelHandlerContext) {
+            if (repetitionsRemaining > 0) {
+                do {
+                    repetitionsRemaining -= 1
+                    
+                    // Get the server address.
+                    let remoteAddress = try self.remoteAddressInitializer()
+                    
+                    // Set the transmission data.
+                    let line = "Something to send there and back again."
+                    var buffer = context.channel.allocator.buffer(capacity: line.utf8.count)
+                    buffer.writeString(line)
+                    self.numBytesOutstanding += buffer.readableBytes
+                    
+                    // Forward the data.
+                    let envolope = AddressedEnvelope<ByteBuffer>(remoteAddress: remoteAddress, data: buffer)
+                    
+                    context.writeAndFlush(self.wrapOutboundOut(envolope), promise: nil)
+                    
+                } catch {
+                    print("Could not resolve remote address")
+                }
+            }
+            else {
+                // We're all done - hurrah!
+                context.close(promise: nil)
+            }
+        }
+        
+        public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+            let envelope = self.unwrapInboundIn(data)
+            let byteBuffer = envelope.data
+            
+            self.numBytesOutstanding -= byteBuffer.readableBytes
+            
+            if self.numBytesOutstanding <= 0 {
+                // Got back everything we sent - maybe send some more.
+                sendSomeDataIfDesiredOrClose(context: context)
+            }
+        }
+        
+        public func errorCaught(context: ChannelHandlerContext, error: Error) {
+            print("error: ", error)
+            
+            // As we are not really interested getting notified on success or failure we just pass nil as promise to
+            // reduce allocations.
+            context.close(promise: nil)
+        }
+    }
+
+    static func doUdpRequests(group: EventLoopGroup, number numberOfRequests: Int) throws -> Int {
+        let serverChannel = try DatagramBootstrap(group: group)
+            .channelOption(ChannelOptions.socketOption(.reuseaddr), value: 1)
+            // Set the handlers that are applied to the bound channel
+            .channelInitializer { channel in
+                // Ensure we don't read faster than we can write by adding the BackPressureHandler into the pipeline.
+                return channel.pipeline.addHandler(EchoHandler())
+            }
+            .bind(to: localhostPickPort).wait()
+
+        defer {
+            try! serverChannel.close().wait()
+        }
+        
+        let remoteAddress = { () -> SocketAddress in serverChannel.localAddress! }
```

Comment:
Why is this a closure, instead of just passing it directly?

Reply:
As above - cloned from existing example - will try without.

Reply:
Made direct and it worked.